File: 3-webservices/REST_Songs_AJAX/src/main/webapp/SongResource.js
/* 
 * Copyright © 2018 Dennis Schulmeister-Zimolong
 * 
 * E-Mail: dhbw@windows3.de
 * Webseite: https://www.wpvs.de/
 * 
 * Dieser Quellcode ist lizenziert unter einer
 * Creative Commons Namensnennung 4.0 International Lizenz.
 */
"use strict";

/**
 * Von der Klasse SongResource des Servers abgeleitete Klasse, die im Prinzip
 * dieselben Methoden besitzt. Hier rufen wir jedoch den REST-Webservice des
 * Servers auf, anstelle direkt auf eine Datenbank zuzugreifen.
 */
class SongResource {

    /**
     * Konstruktor.
     * @param {String} url Basis-URL des REST-Webservices (optional)
     */
    constructor(url) {
        this.url = url || "http://localhost:8080/REST_Songs_Server/api/Songs/";
        this.username = "";
        this.password = "";
    }

    /**
     * Benutzername und Passwort für die Authentifizierung merken.
     * @param {String} username Benutzername
     * @param {String} password Passwort
     */
    setAuthData(username, password) {
        this.username = username;
        this.password = password;
    }

    /**
     * Songs suchen.
     * @param {String} query Suchparameter (optional)
     * @returns {Promise} Gefundene Songs
     */
    async findSongs(query) {
        let url = this.url;

        if (query !== undefined) {
            url += "?query=" + encodeURI(query);
        }

        let response = await fetch(url, {
            headers: {
                "accept": "application/json"
            }
        });

        return await response.json();
    }

    /**
     * Neuen Song speichern.
     * @param {Object} song Zu speichernder Song
     * @returns {Promise} Gespeicherter Song
     */
    async saveNewSong(song) {
        let response = await fetch(this.url, {
            method: "POST",
            headers: {
                "accept": "application/json",
                "content-type": "application/json",
                "authorization": "Basic " + btoa(this.username + ":" + this.password)
            },
            body: JSON.stringify(song)
        });

        return await response.json();
    }

    /**
     * Einzelnen Song auslesen.
     * @param {Number} id Song-ID
     * @returns {Promise} Gefundener Song
     */
    async getSong(id) {
        let response = await fetch(this.url + id + "/", {
            headers: {
                "accept": "application/json",
                "authorization": "Basic " + btoa(this.username + ":" + this.password)
            }
        });

        return await response.json();
    }

    /**
     * Aktualisieren eines Songs.
     * @param {Object} song Zu speichernder Song
     * @returns {Promise} Gespeicherter Song
     */
    async updateSong(song) {
        let response = await fetch(this.url + song.id + "/", {
            method: "POST",
            headers: {
                "accept": "application/json",
                "content-type": "application/json",
                "authorization": "Basic " + btoa(this.username + ":" + this.password)
            },
            body: JSON.stringify(song)
        });

        return await response.json();
    }

    /**
     * Song löschen.
     * @param {Number} id Song ID
     * @returns {Promise} Gelöschter Song
     */
    async deleteSong(id) {
        let response = await fetch(this.url + id + "/", {
            method: "DELETE",
            headers: {
                "accept": "application/json",
                "authorization": "Basic " + btoa(this.username + ":" + this.password)
            }
        });

        return await response.json();
    }
}

File: 3-webservices/REST_Songs_Client/src/main/java/dhbwka/wwi/vertsys/ws/rest_songs_client/SongResource.java
/*
 * Copyright © 2018 Dennis Schulmeister-Zimolong
 * 
 * E-Mail: dhbw@windows3.de
 * Webseite: https://www.wpvs.de/
 * 
 * Dieser Quellcode ist lizenziert unter einer
 * Creative Commons Namensnennung 4.0 International Lizenz.
 */
package dhbwka.wwi.vertsys.ws.rest_songs_client;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.Unirest;
import com.mashape.unirest.http.exceptions.UnirestException;

/**
 * Von der Klasse SongResource des Servers abgeleitete Klasse, die im Prinzip
 * dieselben Methoden besitzt. Hier rufen wir jedoch den REST-Webservice des
 * Servers auf, anstelle direkt auf eine Datenbank zuzugreifen.
 */
public class SongResource {

    public static final String URL = "http://localhost:8080/REST_Songs_Server/api/Songs/";

    public String url = URL;
    public String username = "";
    public String password = "";

    Gson gson = new GsonBuilder().create();

    //<editor-fold defaultstate="collapsed" desc="Konstruktoren">
    public SongResource() {
    }

    public SongResource(String url) {
        this.url = url;
    }
    //</editor-fold>

    /**
     * Benutzername und Passwort für die Authentifizierung merken.
     *
     * @param username Benutzername
     * @param password Passwort
     */
    public void setAuthData(String username, String password) {
        this.username = username;
        this.password = password;
    }
    
    /**
     * Songs suchen.
     * 
     * @param query Suchbegriff
     * @return Gefundene Songs
     * @throws UnirestException HTTP-Fehler
     * @throws WebAppException Server-Fehler
     */
    public Song[] findSongs(String query) throws UnirestException, WebAppException {
        HttpResponse<String> httpResponse = Unirest.get(this.url)
                .queryString("query", query)
                .header("accept", "application/json")
                .asString();
        
        try {
            ExceptionResponse er = this.gson.fromJson(httpResponse.getBody(), ExceptionResponse.class);
            
            if (er.exception != null) {
                throw new WebAppException(er);
            }
        } catch (JsonSyntaxException ex) {
            // Okay, keine Exception empfangen
        }
        
        return this.gson.fromJson(httpResponse.getBody(), Song[].class);
    }
    
    /**
     * Neuen Song speichern.
     * 
     * @param song Zu speichernder Song
     * @return Gespeicherter Song
     * @throws UnirestException HTTP-Fehler
     * @throws WebAppException Server-Fehler
     */
    public Song saveNewSong(Song song) throws UnirestException, WebAppException {
        HttpResponse<String> httpResponse = Unirest.post(this.url)
                .header("accept", "application/json")
                .header("content-type", "application/json")
                .basicAuth(this.username, this.password)
                .body(this.gson.toJson(song))
                .asString();
        
        try {
            ExceptionResponse er = this.gson.fromJson(httpResponse.getBody(), ExceptionResponse.class);
            
            if (er.exception != null) {
                throw new WebAppException(er);
            }
        } catch (JsonSyntaxException ex) {
            // Okay, keine Exception empfangen
        }
        
        return this.gson.fromJson(httpResponse.getBody(), Song.class);
    }
    
    /**
     * Einzelnen Song auslesen.
     * 
     * @param id Song-ID
     * @return Gefundener Song
     * @throws UnirestException HTTP-Fehler
     * @throws WebAppException Server-Fehler
     */
    public Song getSong(long id) throws UnirestException, WebAppException {
        HttpResponse<String> httpResponse = Unirest.get(this.url + id + "/")
                .header("accept", "application/json")
                .asString();
        
        try {
            ExceptionResponse er = this.gson.fromJson(httpResponse.getBody(), ExceptionResponse.class);
            
            if (er.exception != null) {
                throw new WebAppException(er);
            }
        } catch (JsonSyntaxException ex) {
            // Okay, keine Exception empfangen
        }
        
        return this.gson.fromJson(httpResponse.getBody(), Song.class);
    }
    
    /**
     * Aktualisieren eines Songs.
     * 
     * @param song Zu speichernder Song
     * @return Gespeicherter Song
     * @throws UnirestException HTTP-Fehler
     * @throws WebAppException Server-Fehler
     */
    public Song updateSong(Song song) throws UnirestException, WebAppException {
        HttpResponse<String> httpResponse = Unirest.post(this.url + song.id + "/")
                .header("accept", "application/json")
                .header("content-type", "application/json")
                .basicAuth(this.username, this.password)
                .body(this.gson.toJson(song))
                .asString();
        
        try {
            ExceptionResponse er = this.gson.fromJson(httpResponse.getBody(), ExceptionResponse.class);
            
            if (er.exception != null) {
                throw new WebAppException(er);
            }
        } catch (JsonSyntaxException ex) {
            // Okay, keine Exception empfangen
        }
        
        return this.gson.fromJson(httpResponse.getBody(), Song.class);
    }
    
    /**
     * Song löschen.
     * 
     * @param id Song-ID
     * @return Gelöschter Song
     * @throws UnirestException HTTP-Fehler
     * @throws WebAppException Server-Fehler
     */
    public Song deleteSong(long id) throws UnirestException, WebAppException {
        HttpResponse<String> httpResponse = Unirest.delete(this.url + id + "/")
                .header("accept", "application/json")
                .basicAuth(this.username, this.password)
                .asString();
        
        try {
            ExceptionResponse er = this.gson.fromJson(httpResponse.getBody(), ExceptionResponse.class);
            
            if (er.exception != null) {
                throw new WebAppException(er);
            }
        } catch (JsonSyntaxException ex) {
            // Okay, keine Exception empfangen
        }
        
        return this.gson.fromJson(httpResponse.getBody(), Song.class);
    }
    
}
